metalearners/grid_search.py

```diff
@@ -0,0 +1,300 @@
+    start_time = time.time()
+    job.metalearner.fit(job.X_train, job.y_train, job.w_train, **job.kwargs)
+    fit_time = time.time() - start_time
+
+    train_scores = job.metalearner.evaluate(
+        X=job.X_train,
+        y=job.y_train,
+        w=job.w_train,
+        is_oos=False,
+        scoring=job.scoring,
+    )
+    if job.X_test is not None and job.y_test is not None and job.w_test is not None:
+        test_scores = job.metalearner.evaluate(
+            X=job.X_test,
+            y=job.y_test,
+            w=job.w_test,
+            is_oos=True,
+            oos_method=job.oos_method,
+            scoring=job.scoring,
+        )
+    else:
+        test_scores = None
+    score_time = time.time() - fit_time
+    return _GSResult(
+        metalearner=job.metalearner,
+        fit_time=fit_time,
+        score_time=score_time,
+        train_scores=train_scores,
+        test_scores=test_scores,
+    )
+
+
+def _format_results(results: Sequence[_GSResult]) -> pd.DataFrame:
+    rows = []
+    for result in results:
+        row: dict[str, str | int | float] = {}
+        row["metalearner"] = result.metalearner.__class__.__name__
+        nuisance_models = (
+            set(result.metalearner.nuisance_model_specifications().keys())
+            - result.metalearner._prefitted_nuisance_models
+        )
+        treatment_models = set(
+            result.metalearner.treatment_model_specifications().keys()
+        )
+        for model_kind in nuisance_models:
+            row[model_kind] = result.metalearner.nuisance_model_factory[
+                model_kind
+            ].__name__
+            for param, value in result.metalearner.nuisance_model_params[
+                model_kind
+            ].items():
+                row[f"{model_kind}_{param}"] = value
+        for model_kind in treatment_models:
+            row[model_kind] = result.metalearner.treatment_model_factory[
+                model_kind
+            ].__name__
+            for param, value in result.metalearner.treatment_model_params[
+                model_kind
+            ].items():
+                row[f"{model_kind}_{param}"] = value
+        row["fit_time"] = result.fit_time
+        row["score_time"] = result.score_time
+        for name, value in result.train_scores.items():
+            row[f"train_{name}"] = value
+        if result.test_scores is not None:
+            for name, value in result.test_scores.items():
+                row[f"test_{name}"] = value
+        rows.append(row)
+    df = pd.DataFrame(rows)
+    return df
+
+
+class MetaLearnerGridSearch:
+    """Exhaustive search over specified parameter values for a MetaLearner.
+
+    ``metalearner_params`` should contain the necessary params for the MetaLearner initialization
+    such as ``n_variants`` and ``is_classification``. It can also contain optional parameters
+    that all MetaLearners should be initialized with such as ``n_folds`` or ``feature_set``.
```

Comment:
```suggestion
    that all MetaLearners can be initialized with such as ``n_folds`` or ``feature_set``.
```

Reply:
I am unsure about this, check [this](https://github.com/Quantco/metalearners/pull/9#discussion_r1656589419). Lmk if further sth is not clear.

Reply:
Hmm I think I understand your point but this way of phrasing it doesn't seem perfectly obvious to me. What about

If one wants to pass optional parameters to the ``MetaLearners`` initialization, such as ``n_folds`` or ``feature_set`` this should be done by this way, too.

Reply:
[669f37f](https://github.com/Quantco/metalearners/pull/9/commits/669f37f83c034bce59506bd82194f7868db7e4ce)